main.go

```diff
@@ -410,6 +536,116 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
 	ch <- prometheus.MustNewConstMetric(e.virtualMemory, prometheus.CounterValue, parse(s, "ps_vsize"))
 
 	ch <- prometheus.MustNewConstMetric(e.asynclogRequests, prometheus.CounterValue, parse(s, "asynclog_requests"))
+
+	if e.server_stats {
+		// Per-server stats
+		s1, err := getServerStats(conn)
+
+		if err != nil {
+			ch <- prometheus.MustNewConstMetric(e.up, prometheus.GaugeValue, 0)
+			log.Errorf("Failed to collect server stats from mcrouter: %s", err)
+			return
+		}
+
+		for server, metrics := range s1 {
+			ch <- prometheus.MustNewConstMetric(
+				e.serverDuration, prometheus.GaugeValue, parse(metrics, "avg_latency_us"), server)
+			ch <- prometheus.MustNewConstMetric(
+				e.serverProxyReqsProcessing, prometheus.GaugeValue, parse(metrics, "pending_reqs"), server)
+			ch <- prometheus.MustNewConstMetric(
+				e.serverProxyInflightReqs, prometheus.GaugeValue, parse(metrics, "inflight_reqs"), server)
+			ch <- prometheus.MustNewConstMetric(
+				e.serverProxyRetransRatio, prometheus.GaugeValue, parse(metrics, "avg_retrans_ratio"), server)
+
+			_, ok := metrics["stored"]
+			server_stored := 0.0
+			if ok {
+				server_stored = parse(metrics, "stored")
+			}
+			ch <- prometheus.MustNewConstMetric(
+				e.serverMemcachedStored, prometheus.CounterValue, server_stored, server)
+
+			_, ok = metrics["not_stored"]
+			server_not_stored := 0.0
+			if ok {
+				server_not_stored = parse(metrics, "not_stored")
+			}
+			ch <- prometheus.MustNewConstMetric(
+				e.serverMemcachedNotStored, prometheus.CounterValue, server_not_stored, server)
+
+			_, ok = metrics["found"]
+			server_found := 0.0
+			if ok {
+				server_found = parse(metrics, "found")
+			}
+			ch <- prometheus.MustNewConstMetric(
+				e.serverMemcachedFound, prometheus.CounterValue, server_found, server)
+
+			_, ok = metrics["not_found"]
+			server_not_found := 0.0
+			if ok {
+				server_not_found = parse(metrics, "not_found")
+			}
+			ch <- prometheus.MustNewConstMetric(
+				e.serverMemcachedNotFound, prometheus.CounterValue, server_not_found, server)
+
+			_, ok = metrics["deleted"]
+			server_deleted := 0.0
+			if ok {
+				server_deleted = parse(metrics, "deleted")
+			}
+			ch <- prometheus.MustNewConstMetric(
+				e.serverMemcachedDeleted, prometheus.CounterValue, server_deleted, server)
+
+			_, ok = metrics["touched"]
+			server_touched := 0.0
+			if ok {
+				server_touched = parse(metrics, "touched")
+			}
+			ch <- prometheus.MustNewConstMetric(
+				e.serverMemcachedTouched, prometheus.CounterValue, server_touched, server)
+
+			_, ok = metrics["exists"]
+			server_exists := 0.0
+			if ok {
+				server_exists = parse(metrics, "exists")
```

Comment:
Rather then having a lot of duplication here, i think we should extend `parse` or provide a new helper function to default the value to `0.0` if the key doesn't exist in the map.

Ends up removing a lot of the default value/map checking code in this block e.g.

```go
func parse(stats map[string]string, key string) float64 {
  val, ok = metrics[key]  
  if !ok {
    return 0.0
  }
  v, err := strconv.ParseFloat(val, 64)
...
```

```go
      server_not_found := parse(metrics, "not_found")
      ch <- prometheus.MustNewConstMetric(
        e.serverMemcachedNotFound, prometheus.CounterValue, server_not_found, server)
```

Reply:
Really elegant, didn't think about it, will fix!